### PR TITLE
Create chain dependent state hook for fast chain switching

### PR DESCRIPTION
**Pull request by haydenshively**
Basically just keeps track of which chain a given piece of state belongs to so that we don't have to refresh the page. In theory we could use cached data and not have to re-fetch anything after a chain has been loaded once, but right now it'll still refetch (_not_ refresh) every time you switch chains.

Expect a few contractRead errors on the Portfolio Page due to the modals referencing stale data (tokens that exist on one chain but not another). They're ugly in the console but it figures itself out.

Also, I counted the number of renders before and after this change and it's the same (or +1 in some cases). So I think it's safe to use.

We could merge this to master if I take out the BoostPage changes.

### CODE REVIEW

File: shared/src/components/navbar/ChainSelector.tsx [hunk 1]
```diff
@@ -49,20 +48,21 @@ export default function ChainSelector(props: ChainSelectorProps) {
...
-    onSuccess: () => {
+    onMutate(args) {
+      console.debug('About to switch chains', args);
+    },
+    onSuccess: (chain) => {
+      setActiveChain(chain);
       setShouldAttemptToSwitchNetwork(false);
       setIsOpen(false);
       setPendingChainOption(undefined);
-      navigate(0);
```
Comment (haydenshively) [inline shared/src/components/navbar/ChainSelector.tsx:65]: Taking this out is what gives us the speed

File: shared/src/data/hooks/UseChainDependentState.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { useCallback, useMemo, useRef, useState } from 'react';
+
+export function useChainDependentState<S>(initialState: S, chainId: number) {
```
Comment (haydenshively) [inline shared/src/data/hooks/UseChainDependentState.ts:3]: Drop-in replacement for React's `useState`

File: earn/src/components/portfolio/modal/WithdrawModal.tsx [hunk 3]
```diff
@@ -83,9 +83,8 @@ export default function WithdrawModal(props: WithdrawModalProps) {
   const account = useAccount();
 
   const filteredPairs = lendingPairs.filter((p) => doesLendingPairContainToken(p, selectedToken));
-  if (filteredPairs.length === 0) throw new Error(`${selectedToken.symbol} isn't part of any lending pair`);
-  const selectedPair = filteredPairs.at(selectedPairIdx) ?? filteredPairs[0];
-  const lender = selectedPair.token0.equals(selectedToken) ? selectedPair.kitty0 : selectedPair.kitty1;
+  const selectedPair = filteredPairs.at(selectedPairIdx);
+  const lender = selectedPair?.token0.equals(selectedToken) ? selectedPair.kitty0 : selectedPair?.kitty1;
   const amount = GN.fromDecimalString(inputValue[0] || '0', selectedToken.decimals);
```
Comment (haydenshively) [inline earn/src/components/portfolio/modal/WithdrawModal.tsx:88]: Ideally wouldn't have had to change this file, but there was some bad code in here from last time I touched it

File: earn/src/pages/BoostPage.tsx [hunk 4]
```diff
@@ -26,10 +27,60 @@ export default function BoostPage() {
...
+            DEFAULT_COLOR1,
+            res.uniswapPosition,
+            res.uniswapFees,
+            res.borrower
+          );
+        })
+      );
+
+      if (mounted) setBoostedCardInfos(fetchedInfos);
+    })();
```
Comment (haydenshively) [inline earn/src/pages/BoostPage.tsx:71]: Tbh idk why this is showing up on this diff. Should've already been on master.

**Review by IanWoodard — APPROVED** [review]

**Pull request merged**